### PR TITLE
[quant] Add intrinsic module mappings

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23753 [quant] Add intrinsic module mappings**

Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D16820749](https://our.internmc.facebook.com/intern/diff/D16820749)

### CODE REVIEW

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Can you please add a test that checks for module swapping for the newly added entries in the default_module_mapping table?

File: test/common_quantization.py [hunk 1]
```diff
@@ -51,7 +52,6 @@ def test_only_train_fn(model, train_data, loss_fn=_default_loss_fn):
             optimizer.zero_grad()
             output = model(data)
             loss = loss_fn(output, target)
-            loss.backward()
```
Comment (raghuramank100) [inline test/common_quantization.py]: Why is this line removed? Without it backprop will not occur.

Comment (jerryzh168) [reply test/common_quantization.py]: oh, removed when debugging, will add back

File: test/test_qat.py [hunk 2]
```diff
@@ -129,9 +129,8 @@ def ref_op(x):
 
             # backward
             dout = torch.randn(result_ref.size(), dtype=torch.float)
-            result_actual.backward(dout, retain_graph=True)
-            grad_ref = input.grad.cpu()
             result_actual.backward(dout)
+            grad_ref = input.grad.cpu()
```
Comment (raghuramank100) [inline test/test_qat.py]: Gradient calculation for the reference is missing. Suggest:
```python
result_ref = ref_op(input)
result_actual = qat_op(input)
dout = torch.randn(result_ref.size(), dtype=torch.float)
result_ref.backward(dout)
grad_ref = input.grad.cpu()
input.grad.data.zero_()
result_actual.backward(dout)
grad_actual = input.grad.cpu()
self.assertEqual(grad_ref, grad_actual)
```

w1.grad.data.zero_()
result_ref.backward(dout)

File: test/common_quantization.py [hunk 3]
```diff
@@ -90,7 +91,7 @@ def checkObservers(self, module):
             have observers in preperation for quantization
         """
         if hasattr(module, 'qconfig') and module.qconfig is not None and len(module._modules) == 0:
-            self.assertTrue(hasattr(module, 'observer'))
+            self.assertTrue(hasattr(module, 'observer'), 'module: ' + str(type(module)) + ' do not have observer')
```
Comment (zafartahirov) [inline test/common_quantization.py]: 80 char limit?

```python
self.assertTrue(hasattr(module, 'observer'), \
    'module: ' + str(type(module)) + ' do not have observer')
```

File: test/common_quantization.py [hunk 4]
```diff
@@ -352,30 +353,50 @@ def forward(self, x):
...
 
 
-class SubModForFusion(torch.nn.Module):
+class SubModelForFusion(nn.Module):
     def __init__(self):
-        super(SubModForFusion, self).__init__()
-        self.conv = torch.nn.Conv2d(20, 20, 1, bias=None)
-        self.bn = torch.nn.BatchNorm2d(20)
+        super(SubModelForFusion, self).__init__()
+        self.conv = nn.Conv2d(2, 2, 1, bias=None).to(dtype=torch.float)
```
Comment (zafartahirov) [inline test/common_quantization.py]: Why do we need to convert the layer to `float`? I think by default it assumes so.

File: torch/nn/_intrinsic/qat/modules/conv_fused.py [hunk 5]
```diff
@@ -150,11 +148,11 @@ def from_float(cls, mod, qconfig=None):
...
             or directly from user
         """
-        assert type(mod) == cls.__FLOAT_MODULE, 'qat.' + cls.__name__ + '.from_float only works for ' + \
-            cls.__FLOAT_MODULE.__name__
+        assert type(mod) == cls._FLOAT_MODULE, 'qat.' + cls.__name__ + '.from_float only works for ' + \
+            cls._FLOAT_MODULE.__name__
         if not qconfig:
             assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
-            assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
+            assert mod.qconfig, 'Input float module must has valid qconfig'
```
Comment (zafartahirov) [inline torch/nn/_intrinsic/qat/modules/conv_fused.py]: nit: `have valid qconfig`

File: torch/nn/_intrinsic/quantized/modules/conv_relu.py [hunk 6]
```diff
@@ -22,6 +24,18 @@ def __init__(self, in_channels, out_channels, kernel_size, stride=1,
                                          stride=stride, padding=padding, dilation=dilation,
                                          groups=groups, bias=bias, padding_mode=padding_mode)
 
+    @property
+    def weight(self):
```
Comment (zafartahirov) [inline torch/nn/_intrinsic/quantized/modules/conv_relu.py]: Did we decide to keep properties for the weight?

File: torch/nn/utils/fusion.py [hunk 7]
```diff
@@ -8,22 +8,16 @@ def fuse_conv_bn_eval(conv, bn):
...
-    if b_conv is None:
-        b_conv = bn_mean.new_zeros(bn_mean.shape)
+    return fused_conv
 
-    w_conv = w_conv * (bn_weight / bn_var_sqrt).reshape([-1, 1, 1, 1])
-    b_conv = (b_conv - bn_mean) / bn_var_sqrt * bn_weight + bn_bias
+def fuse_conv_bn_weights(conv_w, conv_b, bn_rm, bn_rv, bn_eps, bn_w, bn_b):
+    if conv_b is None:
+        conv_b = bn_rm.new_zeros(bn_rm.shape)
+    bn_var_sqrt = torch.sqrt(bn_rv + bn_eps)
```
Comment (zafartahirov) [inline torch/nn/utils/fusion.py]: nit: `bn_var_sqrt = torch.rsqrt(bn_rv + bn_eps)`

File: torch/nn/utils/fusion.py [hunk 8]
```diff
@@ -8,22 +8,16 @@ def fuse_conv_bn_eval(conv, bn):
...
 
-    fused_conv.weight = torch.nn.Parameter(w_conv)
-    fused_conv.bias = torch.nn.Parameter(b_conv)
+    conv_w = conv_w * (bn_w / bn_var_sqrt).reshape([-1, 1, 1, 1])
+    conv_b = (conv_b - bn_rm) / bn_var_sqrt * bn_w + bn_b
```
Comment (zafartahirov) [inline torch/nn/utils/fusion.py]: nit: If using `rsqrt` above, you can rewrite to

```python
    conv_w = conv_w * (bn_w * bn_var_rsqrt).reshape([-1, 1, 1, 1])
    conv_b = (conv_b - bn_rm) * bn_var_rsqrt * bn_w + bn_b
```

**Review by z-a-f — APPROVED** [review]
Minor nits

File: test/common_quantization.py (hunk 4 above)
Comment (jerryzh168) [reply test/common_quantization.py]: by default it is torch.double

File: torch/nn/_intrinsic/quantized/modules/conv_relu.py (hunk 6 above)
Comment (jerryzh168) [reply torch/nn/_intrinsic/quantized/modules/conv_relu.py]: no, these should be removed

Comment (jerryzh168) [reply torch/nn/_intrinsic/quantized/modules/conv_relu.py]: this pr already has too many changes, will do it in a separate pr

**Comment by ezyang** [comment]
@jerryzh168 Hi Jerry, I'm not too familiar with the workstream here. What kind of review are you looking for?

File: torch/nn/_intrinsic/qat/modules/conv_fused.py [hunk 9]
```diff
@@ -129,8 +127,8 @@ def _forward(self, input):
             conv = conv * rescale_factor.reshape([1, -1, 1, 1])
             conv = conv + (self.beta - self.gamma * batch_mean * batch_rstd).reshape([1, -1, 1, 1])
 
-            self.running_mean = exponential_average_factor * batch_mean + (1 - exponential_average_factor) * self.running_mean
-            self.running_var = exponential_average_factor * batch_var + (1 - exponential_average_factor) * self.running_var
+            self.running_mean = exponential_average_factor * batch_mean.detach() + (1 - exponential_average_factor) * self.running_mean
```
Comment (jerryzh168) [inline torch/nn/_intrinsic/qat/modules/conv_fused.py]: @ezyang just want to make sure these detach() are used correctly.

Comment (ezyang) [reply torch/nn/_intrinsic/qat/modules/conv_fused.py]: LGTM

File: test/test_qat.py [hunk 10]
```diff
@@ -123,17 +122,21 @@ def ref_op(x):
             else:
                 ref_op = compose([conv_op, bn_op, relu_op])
 
-            result_ref = ref_op(input)
-            result_actual = qat_op(input)
-            self.assertEqual(result_ref, result_actual)
+            input_clone = input.clone().detach_().requires_grad_()
+            for i in range(2):
```
Comment (raghuramank100) [inline test/test_qat.py]: This looks like a typo. Do you need the for loop here?

File: test/test_qat.py [hunk 11]
```diff
@@ -123,17 +122,21 @@ def ref_op(x):
...
-            grad_actual = input.grad.cpu()
-            self.assertEqual(grad_ref, grad_actual)
+                # backward
+                dout = torch.randn(result_ref.size(), dtype=torch.float)
+                loss = (result_ref - dout).sum()
+                loss.backward()
+                grad_ref = input.grad.cpu()
+                loss = (result_actual - dout).sum()
+                loss.backward()
+                grad_actual = input_clone.grad.cpu()
```
Comment (raghuramank100) [inline test/test_qat.py]: NIT: You dont need to create input_clone, you could set gradient to zero after you calculate grad_ref.
```
input.grad.zero_()
```
would do that.

File: torch/nn/qat/modules/linear.py [hunk 12]
```diff
@@ -39,11 +39,11 @@ def from_float(cls, mod, qconfig=None):
...
             or directly from user
         """
-        assert type(mod) == cls.__FLOAT_MODULE, ' qat.' + cls.__name__ + '.from_float only works for ' + \
-            cls.__FLOAT_MODULE.__name__
+        assert type(mod) == cls._FLOAT_MODULE, ' qat.' + cls.__name__ + '.from_float only works for ' + \
+            cls._FLOAT_MODULE.__name__
         if not qconfig:
             assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
-            assert mod.qconfig, 'Input float module must has valid qconfig'
+            assert mod.qconfig, 'Input float module must have a valid qconfig'
```
Comment (raghuramank100) [inline torch/nn/qat/modules/linear.py]: Do we need this check after  assert hasattr(mod, 'qconfig'). Is the second check ensuring that qconfig is not None?

File: torch/nn/_intrinsic/quantized/modules/conv_relu.py (hunk 6 above)
Comment (raghuramank100) [reply torch/nn/_intrinsic/quantized/modules/conv_relu.py]: If you are inheriting from nn.quantized.conv2d, with the latest changes, you dont need these functions anymore, delete these lines?

File: torch/nn/_intrinsic/quantized/modules/conv_relu.py [hunk 13]
```diff
@@ -33,3 +48,11 @@ def forward(self, input):
...
                                                         float(self.scale), int(self.zero_point))
         return output.permute([0, 3, 1, 2])
+
+    @classmethod
+    def from_float(cls, mod):
+        if type(mod) == nniqat.ConvBnReLU2d:
+            mod.weight, mod.bias = \
+                fuse_conv_bn_weights(mod.weight, mod.bias, mod.running_mean,
+                                     mod.running_var, mod.eps, mod.gamma, mod.beta)
+        return super(ConvReLU2d, cls).from_float(mod)
```
Comment (raghuramank100) [inline torch/nn/_intrinsic/quantized/modules/conv_relu.py]: Wouldn't super return a nn.quantized.conv2d module instead of a nn._intrinsic.quantized.conv2drelu?

File: test/test_qat.py (hunk 10 above)
Comment (jerryzh168) [reply test/test_qat.py]: yeah, I had some error previously when I run the op with forward and backward twice

File: test/test_qat.py (hunk 11 above)
Comment (jerryzh168) [reply test/test_qat.py]: in that case we still need to store the gradient of input somewhere before we run a different graph.

File: torch/nn/qat/modules/linear.py (hunk 12 above)
Comment (jerryzh168) [reply torch/nn/qat/modules/linear.py]: yeah this is to make sure we have a valid qconfig

File: torch/nn/_intrinsic/quantized/modules/conv_relu.py (hunk 13 above)
Comment (jerryzh168) [reply torch/nn/_intrinsic/quantized/modules/conv_relu.py]: no, we are using cls in the from_float

Comment (raghuramank100) [reply torch/nn/_intrinsic/quantized/modules/conv_relu.py]: I see, cls is pointing to conv2drelu

File: torch/nn/quantized/modules/conv.py [hunk 14]
```diff
@@ -126,14 +129,21 @@ def from_float(cls, mod):
...
-                cls.__FLOAT_MODULE.__name__
+            assert type(mod) == cls._FLOAT_MODULE, ' nnq.' + cls.__name__ + '.from_float only works for ' + \
+                cls._FLOAT_MODULE.__name__
             assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
-            assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
             weight_observer = mod.qconfig.weight()
+            # workaround for sequential, ConvReLU2d should probably
+            # inherit from Conv2d instead
+            if type(mod) == nni.ConvReLU2d:
+                mod = mod[0]
```
Comment (raghuramank100) [inline torch/nn/quantized/modules/conv.py]: We have a problem when we deal with nn._intrinsic.convReLU2d. We want to use the observer after the ReLU, but here we end up using the observer attached to the conv.

File: torch/nn/quantized/modules/linear.py [hunk 15]
```diff
@@ -196,10 +199,15 @@ def from_float(mod):
...
-            assert type(mod) == NNLinear, 'nnq.Linear.from_float only works for nn.Linear'
+            assert type(mod) == cls._FLOAT_MODULE, ' nnq.' + cls.__name__ + '.from_float only works for ' + \
+                cls._FLOAT_MODULE.__name__
             assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
             assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
             weight_observer = mod.qconfig.weight()
+            # workaround for sequential, ConvReLU2d should probably
+            # inherit from Conv2d instead
+            if type(mod) == nni.LinearReLU:
+                mod = mod[0]
```
Comment (raghuramank100) [inline torch/nn/quantized/modules/linear.py]: Similar comments for linear

File: torch/quantization/quantize.py [hunk 16]
```diff
@@ -19,7 +24,9 @@ def propagate_qconfig_helper(module, qconfig_dict, qconfig_parent=None, prefix='
     Return:
         None, module is modified inplace with qconfig attached
     """
-    if not hasattr(module, 'qconfig'):
+    if type(module) in skip_list:
+        assert not hasattr(module, 'qconfig'), str(type(module)) + ' module is in skip list, it should not have qconfig'
```
Comment (raghuramank100) [inline torch/quantization/quantize.py]: This is not a good user experience. Can we ignore qconfig for modules in skip list?

Comment (jerryzh168) [reply torch/quantization/quantize.py]: you mean silently skip these modules?

File: torch/quantization/quantize.py [hunk 17]
```diff
@@ -240,6 +258,11 @@ def convert(module, mapping=DEFAULT_MODULE_MAPPING):
     module_swapped = swap_module(module, mapping)
 
     reassign = {}
+    # TODO(jerryzh): remove after deciding on the impl of
+    # intrinsic moudles
+    if type(module) in [nni.ConvBn2d, nni.ConvBnReLU2d]:
```
Comment (raghuramank100) [inline torch/quantization/quantize.py]: Why only these two modules? What about nni.ConvRelu2d and nni.LinearRelu2d?

Comment (jerryzh168) [reply torch/quantization/quantize.py]: yeah we need these as well

File: test/test_qat.py [hunk 18]
```diff
@@ -123,17 +122,21 @@ def ref_op(x):
...
+                self.assertEqual(grad_ref, grad_actual)
```
Comment (raghuramank100) [inline test/test_qat.py]: Please check gradients for all parameters in addition to the input: weight, bias, gamma, beta. 
In addition, please check if all running mean/variance estimates exactly match after forward.

File: test/test_qat.py [hunk 19]
```diff
@@ -123,17 +122,40 @@ def ref_op(x):
...
+                running_mean_actual = qat_op.running_mean
+                running_var_actual = qat_op.running_var
+                num_batches_tracked_actual = qat_op.num_batches_tracked
+                self.assertEqual(input_grad_ref, input_grad_actual)
+                self.assertEqual(weight_grad_ref, weight_grad_actual, prec=1e-4)
+                self.assertEqual(gamma_grad_ref, gamma_grad_actual, prec=1e-4)
+                self.assertEqual(beta_grad_ref, beta_grad_actual)
+                self.assertEqual(num_batches_tracked_ref, num_batches_tracked_actual)
+                self.assertEqual(running_mean_ref, running_mean_actual)
+                # self.assertEqual(running_var_ref, running_var_actual, prec=1e-2)
```
Comment (raghuramank100) [inline test/test_qat.py]: The problem is due to incorrect calculation of variance. Batch norm code at: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Normalization.cpp#L223 uses unbiased variance estimate, while the convbn code does not:

https://github.com/pytorch/pytorch/blob/3c5f7c2e4656c84a79168be4e0b7edd8d41e5e2e/torch/nn/_intrinsic/qat/modules/conv_fused.py#L122
Please change:
```
          batch_var = torch.var(conv_orig, dim=[0, 2, 3], unbiased=False)
```
To:
```
          batch_var = torch.var(conv_orig, dim=[0, 2, 3], unbiased=True)
```
  and check if the mismatch in running_var is gone.

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Some suggestions to get running_var to match

**Review by raghuramank100 — APPROVED** [review]
Approved, with follow up needed to understand cause of mismatch of gradients for weights and batch norm gamma

**Comment by jerryzh168** [comment]
lint will is fixed by another pr
